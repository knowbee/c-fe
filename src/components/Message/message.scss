.message {
  padding: 0rem 1.5rem;
  width: 70%;
  margin-left: 345px;
  margin-top: 8rem;

  .message-card {
    display: flex;

    div {
      p {
        background-color: #ddd;
        padding: 1rem;
        color: #555;
        width: 15rem;
        margin-bottom: 0.2rem;
      }

      small {
        color: #aaa;
      }
    }
  }

  .message-card-one {
    display: flex;
    justify-content: flex-end;

    div {
      p {
        background-color: #2aa2fa;
        color: #fff;
        padding: 1rem;
        width: 15rem;
        margin-bottom: 0.2rem;
      }

      small {
        float: right;
      }
    }
  }

  .input-container {
    display: flex;
    justify-content: space-between;
    padding: 1rem;
    width: 100%;
    margin-left: -1rem;
    border-left: none;
    border-bottom: none;
    border-right: none;
    border-top: 1px solid #ddd;

    &:focus {
      outline: none;
    }
    input{
      width: 80%;
      border-top: none;
      border-left: none;
      border-bottom: none;
      border-right: none;
    }
  }

  .send-icon {
    position: relative;
    right: 10rem;
    bottom:1rem;
    padding-top: 1rem;
    cursor: pointer;
  }

  .no-chat {
    text-align: center;
  }
}

@media (max-width: 999px) {
  .message {
    border: none;
  }
}