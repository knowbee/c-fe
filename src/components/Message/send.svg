 <svg
  className="send-icon"
          width="20"
          height="20"
          viewBox="0 0 24 25"
          fill="none"
          xmlns="http://www.w3.org/2000/svg"
         
        >
          <path
            d="M23.6954 0.291234C23.5604 0.156908 23.3899 0.0639048 23.2039 0.0231609C23.0179 -0.0175829 22.8241 -0.00437409 22.6454 0.0612345L0.645352 8.06123C0.45562 8.1332 0.292272 8.26118 0.177003 8.42819C0.0617345 8.59519 0 8.79331 0 8.99623C0 9.19916 0.0617345 9.39728 0.177003 9.56428C0.292272 9.73128 0.45562 9.85927 0.645352 9.93123L9.23535 13.3612L15.5754 7.00123L16.9854 8.41123L10.6154 14.7812L14.0554 23.3712C14.1295 23.5573 14.2577 23.7168 14.4235 23.8292C14.5894 23.9415 14.7851 24.0014 14.9854 24.0012C15.1874 23.9971 15.3835 23.9318 15.5478 23.814C15.712 23.6962 15.8367 23.5313 15.9054 23.3412L23.9054 1.34123C23.9735 1.16431 23.9899 0.971665 23.9527 0.785758C23.9156 0.599851 23.8263 0.428345 23.6954 0.291234Z"
            fill="#2AA2FA"
          />
</svg>