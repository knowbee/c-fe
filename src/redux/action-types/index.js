export * from "./registerUserTypes";
export * from "./loginTypes";
export * from "./currentUserTypes";
export * from "./usersTypes";
