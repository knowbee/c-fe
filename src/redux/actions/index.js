export * from "./registerActions";
export * from "./loginActions";
export * from "./currentUserActions";
