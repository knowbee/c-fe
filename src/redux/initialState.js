module.exports = {
  currentUser: {
    message: "",
    isLoggedIn: localStorage.getItem("token") !== null,
    profile: {},
  },
  register: {
    submitting: false,
    message: "",
    email: "",
    name: "",
    password: "",
    errors: {},
  },
  users: {
    fetching: false,
    chats: [],
    errors: {},
  },
  login: {
    submitting: false,
    credentials: {
      email: "",
      password: "",
    },
  },
};
