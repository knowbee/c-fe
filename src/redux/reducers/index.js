import { combineReducers } from "redux";
import register from "./registerReducer";
import login from "./loginReducer";
import currentUser from "./currentUserReducer";
import users from "./usersReducer";

const reducer = combineReducers({
  register,
  login,
  currentUser,
  users,
});
export default reducer;
