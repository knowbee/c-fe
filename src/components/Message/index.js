import React, { useState, useContext, useEffect, useRef } from "react";
import "./message.scss";
import { connect } from "react-redux";
import { sendMessage, fetchMessages } from "../../redux/actions";
import { SocketContext } from "../../context/socket";

import sendIcon from "./send.svg";
import useSound from "use-sound";
import notification from "./message.mp3";
import TimeAgo from "timeago-react";

function Message({ receiver_id, send, messages, fetchMessages }) {
  const [message, setMessage] = useState("");
  const [sender_id, setSenderId] = useState(null);
  const [newMessage, setNewMessage] = useState({});

  const bottomRef = useRef(null);

  const executeScroll = () => bottomRef.current.scrollIntoView();

  const socket = useContext(SocketContext);

  useEffect(() => {}, [newMessage]);

  const [playOn] = useSound(notification, { volume: 0.4 });

  const handleSendMessage = (receiver_id, message, sender_id) => {
    send({ receiver_id, message, sender_id });
    setMessage("");
  };

  const handleEnterKey = (e, receiver_id, message, sender_id) => {
    if (e.charCode === 13) {
      handleSendMessage(receiver_id, message, sender_id);
    }
  };

  useEffect(() => {
    executeScroll();
  }, [messages]);

  useEffect(() => {
    const current_user = JSON.parse(localStorage.getItem("current_user"));
    socket.on("incomingMessage", (message) => {
      if (message.message.receiver_id === current_user.id) {
        playOn();
      }
      setNewMessage(message);
      fetchMessages(message.message.sender_id);
      fetchMessages(message.message.receiver_id);

      localStorage.setItem("new-message", JSON.stringify(message));
    });

    return () => socket.off("messageSent");
  }, [newMessage, socket, fetchMessages]);

  useEffect(() => {
    const current_user = JSON.parse(localStorage.getItem("current_user"));
    setSenderId(current_user.id);
  }, []);

  return (
    <div className="message">
      <div className="chats">
        {receiver_id ? (
          messages.map((message) => {
            if (receiver_id === message.receiver_id) {
              return (
                <div className="message-card-one" key={message.id}>
                  <div>
                    <p>{message.message}</p>
                    <small>
                      <TimeAgo datetime={message.created_at}></TimeAgo>
                    </small>
                  </div>
                </div>
              );
            } else if (receiver_id === message.sender_id) {
              return (
                <div className="message-card" key={message.id}>
                  <div>
                    <p>{message.message}</p>
                    <small>
                      <TimeAgo datetime={message.created_at}></TimeAgo>
                    </small>
                  </div>
                </div>
              );
            }
          })
        ) : (
          <h2 className="no-chat">No chat selected</h2>
        )}
      </div>

      {receiver_id ? (
        <div className="input-container" id="input">
          <input
            type="text"
            placeholder="Type a message"
            value={message}
            onChange={(e) => setMessage(e.target.value)}
            onKeyPress={(e) => {
              handleEnterKey(e, receiver_id, message, sender_id);
            }}
          ></input>
          <img
            className="send-icon"
            src={sendIcon}
            alt="send"
            onClick={() => {
              handleSendMessage(receiver_id, message, sender_id);
              executeScroll();
            }}
          />
        </div>
      ) : null}
      <div ref={bottomRef}></div>
    </div>
  );
}

const mapStateToProps = ({
  chat: { receiver_id, receiver_name, messages },
}) => {
  return { receiver_id, receiver_name, messages };
};

export const mapDispatchToProps = (dispatch) => ({
  send: ({ receiver_id, message, sender_id }) =>
    dispatch(sendMessage({ receiver_id, message, sender_id })),
  fetchMessages: (id) => dispatch(fetchMessages(id)),
});

export default connect(mapStateToProps, mapDispatchToProps)(Message);
