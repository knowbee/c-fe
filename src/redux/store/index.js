import { createStore, applyMiddleware } from "redux";
import thunk from "redux-thunk";
import { composeWithDevTools } from "redux-devtools-extension/developmentOnly";
import initialState from "../initialState";
import reducers from "../reducers/index";

const store = createStore(
  reducers,
  initialState,
  composeWithDevTools(applyMiddleware(thunk))
);

export default store;
