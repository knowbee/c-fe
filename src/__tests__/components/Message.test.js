import React from "react";
import renderer from "react-test-renderer";
import { Provider } from "react-redux";
import { createStore } from "redux";

import "./index.css";
import rootReducer from "../../redux/reducers";
import Message from "../../components/Message";

const store = createStore(rootReducer);

test("Message component loads with default state", () => {
  const component = renderer.create(
    <Provider store={store}>
      <Message />
    </Provider>
  );

  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
