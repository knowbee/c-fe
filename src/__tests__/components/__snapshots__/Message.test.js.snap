// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Message component loads with default state 1`] = `
<div
  className="message"
>
  <div
    className="chats"
  >
    <h2
      className="no-chat"
    >
      No chat selected
    </h2>
  </div>
  <div />
</div>
`;
